### PR TITLE
fixed health floating point error, fungal id vs name confusion

### PR DESCRIPTION
**Pull request by teh60**
- extra health perk multiplies max HP by 1.5 thus taking it off integer values, so added a floor divide to fix BigInt handling
- fungal shifts were confusing with multiple id's mapping to same translated name, added distinction hover and changed lua mod to report both id and translated name

### CODE REVIEW

**Comment by teh60** [comment]
Only a new Noita mod release is required for dev.onlywands.com for commit af18dc0
- Adds a materials.lua for translating irregular material names to translation keys/UI names
- Allows utils.lua to use materials.lua table as a fallback

**Comment by teh60** [comment]
commit 7a8874d is a change to icons.js within JS backend only, no new version number or noita mod release required

**Comment by teh60** [comment]
commit b6c1de0 :
- Added player location icon onto the map preview system (updates just like the preview image every few seconds)
- Added readout for which heaven/hell loop the player is in just like PW/NG+ counter
- Added fungal shift timer readout (time remaining until you can shift again)
    - Thanks `SeroMaster` from dunk's twitch community for suggesting this feature!
- Refactored map/coordinate related code for future map.runfast.stream integration
- Added debug console.log for weird json.parse exception that I ran into but couldnt reproduce and that dunk has also ran into (seemingly has no actual effect but I should be able to fix it if I can reproduce it/get more diagnostics on it)

commits d9da18c and 40cca7d are self explanatory and I forgot to include them in this main commit

requires both a new JS deploy and lua mod release
## Time remaining on fungal shift and map player icon and heaven/hell loop number readout
image

## Ready to Shift
image

**Comment by teh60** [comment]
commit 4670472 only has changes to the site, no new mod release required
- new "Tinker" button added to wands which loads the wand stats/spells into the noita (beta) wand simulator
    - Thanks to `Dochi` from Dunk's discord for this feature suggestion!
    - Currently uses the beta simulator rather than salinecitrine's original version as the original page errors if epilogue 2 new spells are fed into the site while the beta just ignores not-found spells (not all epilogue 2 spells are supported on the beta yet)
- added wand speed multiplier to wand stats
- added missing wand sprites
    - some wand sprites and wand info were stored in data/entities/items/wands rather than data/items_gfx/items/wands
- fixed justification issue on world seed readout
- fixed fungal shift info error in mongoDB vs live update data
    - previously on page load the data would not be properly fetched from mongoDB and require a data update to be sent from the mod causing confusion with Ready To Shift/how long until you can fungal shift again, and this error also caused fungal shift count to read as undefined under the same scenario

**Comment by teh60** [comment]
commit 316d2fb has both site changes and a new release (beta.5->beta.6).  
Not sure if conga plans on releasing their simple spell lock mod so the integration I added for that may be code that always evals to false (change to utils.lua), however the main.js and spellData.js/spellDataApoth.js changes are needed for clarity/accuracy.

**Comment by teh60** [comment]
commit 9e57e0f updates both the site and the mod release (version 1.2.5-beta)
- during the first 5 minutes of gameplay the fungal shift timer would countdown instead of displaying "Ready to Shift"
- fixed an edgecase where a functional chain was being detected as a broken chain
    - thanks to letaali for alerting me of this error, TLDR toxic group->lava, then lava->ominous showed as a broken chain when really this is just toxic group->lava and lava->lava
- fixed previous shifts being mis-cancelled out when a material shifts into itself
    - thanks to !doit and majesticslothofdoom on dunk's discord for noticing and discussing this bug
    - the example they discussed was oil/swamp/peat->acid then oil/swamp/peat->oil which crossed out oil->acid when its still true since the overriding shift is oil->oil (which is now shown as crossed out to differentiate from oil->(something shifted into oil which displays as oil)

**Comment by teh60** [comment]
commit f882b92 updates both the site and the mod release (fixed to version 1.2.4.rc.1 instead of 1.2.5-beta)

Fungal Shifting (code):
- Main calculator code has been completely rewritten by myndzi, my previous code was a mess to put it lightly
    - Thanks to this rewrite being way better, it was quite easy to checkoff these bucket list features: 
- Main calculator code keeps tracks of what shift number each shift is, including group-shift handling for both better outputting and for previous fungal shift states to be accessible/viewable
- Main calculator code also supports tracing why a shift's resulting material is the way it is, at least a basic form of this concept
- Noita LUA mod now outputs shifts based on their shift number so group material shifts can now be differentiated compared to single material shifts
    - backend DB fungal info also updated to support storing shifts per group/iteration

Fungal Shifting (UI/display):
- Fungal shifts are now in a table for better visual clarity and consistency
- Fungal shifts display their shift number, and have lettering for each sub-material within a fungal group
    - Screenshot 2024-07-11 at 7 40 52 AM
- "Show Broken Chains/Overwrites" former switch reworked into "Show Original Shift in First Column" for table view
    - Screenshot 2024-07-11 at 7 50 44 AM

- Fungal shift "Final Result" materials can be hovered to help trace why a broken chain or functioning chain is the way it is (simple explanation of what materials/shifts are responsible for the resulting material)
     - During this hover other relevant shifts' table cells are highlighted
         - Screenshot 2024-07-11 at 7 44 30 AM
     - Also the material tool-tips say the shift number of the shifts responsible for the resulting material
         - Screenshot 2024-07-11 at 7 43 35 AM
- The fungal shift display can take an input for up to what number shift the resulting material should be based on
    - Screenshot 2024-07-11 at 7 42 04 AM

**Comment by teh60** [comment]
commit 9133943 only updates the JS site, no new release bundle required.
- Added map preview support (and linking to map.runfast.stream) for:
    - NG+
    - Nightmare
    - Apotheosis/Apotheosis NG+
    - Purgatory
    - Noitavania/Noitavania NG+
    - Alternate Biomes
    - Note:  the map preview if in a parallel world tries to use east/west map captures but for most of these map-types map.runfast.stream only gives a main world capture with a tiny bit of parallel world capture, so it'll fallback to just showing the preview as if you were in main world, leading to potential confusion on structures (i.e. lava lake) that exist only in main but not in parallel worlds.  
- Added readout to clarify which map is being displayed

File: public/main.js [hunk 1]
```diff
@@ -538,24 +538,18 @@ const mapComp = Vue.component('map-comp', {
...
-            // debug: "hide-input",
         }
     },
     mounted() {
-        if (this.$refs.tooltip) {
-            this.tooltip = Popper.createPopper(this.$refs.slot, this.$refs.tooltip, {
-                placement: 'right',
-                modifiers: [{ name: 'offset', options: { offset: [0, 5] } }],
+        fetch("https://map.runfast.stream/js/tilesources.json")
+            .then(res => res.json())
```
Comment (myndzi) [inline public/main.js]: A good habit here is something like:

```suggestion
            .then(res => (res.ok ? res.json() : Promise.reject(`HTTP ${res.status}: ${res.statusText}`)))
```

Fetch promises can resolve (successfully complete) with a failure (status in 4xx-5xx) or an unexpected result (status in the 1xx or 3xx range), so you've gotta be sure to catch both "successful failures" and failures

File: public/main.js [hunk 2]
```diff
@@ -538,24 +538,18 @@ const mapComp = Vue.component('map-comp', {
...
+            .then(data => {
+                this.mapData = data
+                this.loaded = true
             })
```
Comment (myndzi) [inline public/main.js]: Do something on failure, too with `.catch(err => ....)`, even if it's just displaying some text like "failed to fetch map data"

File: public/main.js [hunk 3]
```diff
@@ -614,12 +609,20 @@ const mapComp = Vue.component('map-comp', {
             } else if (this.mods.list.includes("noitavania")) {
                 mapName = Number(this.mods.ngp) > 0 ? "noitavania-new-game-plus" : "noitavania"
             }
-            let map = mapData[mapName]
+
+            // get map URL and topleft offset
+            const mapValues = this.mapData[mapName]
+            const map = {
+                urls: mapValues.map((x) => `${x.url.slice(0, -4)}_files/`),
```
Comment (myndzi) [inline public/main.js]: You could use something like `${x.url.replace(/\.dzi$/, '')}` to make the intent clearer here. In real code I'd be looking for validation of the (unknown/untrusted) input data, but the value of adding that here is minimal.

**Review by myndzi — APPROVED** [review]
Between repeated live tests on the dev domain and various other convos, I guess it's time to pull the trigger 😅

**Pull request merged**